###  Data Cleaning Assignment

In [3]:
import pandas as pd
import numpy as np

In [4]:
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN',
'londON_StockhOlm',

'Budapest_PaRis', 'Brussels_londOn'],
'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )',
'12. Air France', '"Swiss Air"']})

In [5]:
df.head()

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN           NaN            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis           NaN          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"

In [8]:
df.describe()

FlightNumber
count           3.0
mean        10065.0
std            20.0
min         10045.0
25%         10055.0
50%         10065.0
75%         10075.0
max         10085.0

In [9]:
df.dtypes

From_To          object
FlightNumber    float64
RecentDelays     object
Airline          object
dtype: object

### Question 1
Some values in the the FlightNumber column are missing. These numbers are
meant to increase by 10 with each row so 10055 and 10075 need to be put in
place. Fill in these missing numbers and make the column an integer column
(instead of a float column).

In [11]:
df['FlightNumber'].fillna(value=df['FlightNumber'][0]+10, limit=1, inplace=True)
df['FlightNumber'].fillna(value=df['FlightNumber'][2]+10, limit=1, inplace=True)

In [14]:
df.head()

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"

In [13]:
df['FlightNumber'] = df['FlightNumber'].astype(np.int64)
df['FlightNumber'].dtype

dtype('int64')

### Question 2
The From_To column would be better as two separate columns! Split each
string on the underscore delimiter _ to give a new temporary DataFrame with
the correct values. Assign the correct column names to this temporary
DataFrame.

In [15]:
temp_data = pd.DataFrame(df['From_To'].str.split('_', 1).to_list(), columns = ['From','To'])

### Question 3
Notice how the capitalisation of the city names is all mixed up in this
temporary DataFrame. Standardise the strings so that only the first letter is
uppercase (e.g. "londON" should become "London".)

In [23]:
df['From'] = temp_data['From'].str.capitalize()
df['To'] = temp_data['To'].str.capitalize()

In [24]:
df.head(6)

From         To           From_To  delay_1  delay_2  delay_3  \
0    London      Paris      LoNDon_paris     23.0     47.0      NaN   
1    Madrid      Milan      MAdrid_miLAN      NaN      NaN      NaN   
2    London  Stockholm  londON_StockhOlm     24.0     43.0     87.0   
3  Budapest      Paris    Budapest_PaRis     13.0      NaN      NaN   
4  Brussels     London   Brussels_londOn     67.0     32.0      NaN   

   FlightNumber              Airline  
0         10045               KLM(!)  
1         10055    <Air France> (12)  
2         10065  (British Airways. )  
3         10075       12. Air France  
4         10085          "Swiss Air"

### Question 4
Delete the From_To column from df and attach the temporary DataFrame
from the previous questions.

In [ ]:
df = pd.concat([temp_data, df], axis=1, sort=False)
df

### Question 5
In the RecentDelays column, the values have been entered into the
DataFrame as a list. We would like each first value in its own column, each

second value in its own column, and so on. If there isn't an Nth value, the value
should be NaN.

In [20]:
delays = pd.DataFrame(df['RecentDelays'].to_list(), columns = ['delay_1', 'delay_2', 'delay_3'])
delays

delay_1  delay_2  delay_3
0     23.0     47.0      NaN
1      NaN      NaN      NaN
2     24.0     43.0     87.0
3     13.0      NaN      NaN
4     67.0     32.0      NaN

In [21]:
df.insert(loc = 3, column='delay_1' , value=delays['delay_1'])
df.insert(loc = 4, column='delay_2' , value=delays['delay_2'])
df.insert(loc = 5, column='delay_3' , value=delays['delay_3'])
df

From         To           From_To  delay_1  delay_2  delay_3  \
0    London      Paris      LoNDon_paris     23.0     47.0      NaN   
1    Madrid      Milan      MAdrid_miLAN      NaN      NaN      NaN   
2    London  Stockholm  londON_StockhOlm     24.0     43.0     87.0   
3  Budapest      Paris    Budapest_PaRis     13.0      NaN      NaN   
4  Brussels     London   Brussels_londOn     67.0     32.0      NaN   

   FlightNumber  RecentDelays              Airline  
0         10045      [23, 47]               KLM(!)  
1         10055            []    <Air France> (12)  
2         10065  [24, 43, 87]  (British Airways. )  
3         10075          [13]       12. Air France  
4         10085      [67, 32]          "Swiss Air"

In [22]:
df.drop(['RecentDelays'], inplace=True, axis=1)
df

From         To           From_To  delay_1  delay_2  delay_3  \
0    London      Paris      LoNDon_paris     23.0     47.0      NaN   
1    Madrid      Milan      MAdrid_miLAN      NaN      NaN      NaN   
2    London  Stockholm  londON_StockhOlm     24.0     43.0     87.0   
3  Budapest      Paris    Budapest_PaRis     13.0      NaN      NaN   
4  Brussels     London   Brussels_londOn     67.0     32.0      NaN   

   FlightNumber              Airline  
0         10045               KLM(!)  
1         10055    <Air France> (12)  
2         10065  (British Airways. )  
3         10075       12. Air France  
4         10085          "Swiss Air"